Embedded link enhancements - June 2023 (#451)
codatio/codat-docs: docs/auth-flow/authorize-embedded-link.md

---
title: "Authorize with Embedded Link"
sidebar_label: Overview

> description: "Embed Link in your application with our authorization flow component"

Reviewer: ```suggestion
description: "Embed our auth flow in your application with our low-code component"
```

---
title: "Authorize with Embedded Link"
sidebar_label: Overview
description: "Embed Link in your application with our authorization flow component"
---

> import Tabs from "@theme/Tabs";

Author: Both of these lines are needed for the "tabbed code block" component, under "Embedding the Link application". I have left an open question in that section.

> import Tabs from "@theme/Tabs";

Reviewer 2: also again, it will not be displayed once it's rendered by docusaurus. it's only github preview that does not know what to do with the tabs

![](/img/link/0014-embedded-link-demo.png)

:::caution Embedded Link is in beta

Embedded Link is in beta, with more enhancements coming soon. You can report any issues with the component by contacting [Codat Support](mailto:support@codat.io).

If you have general feedback on the component, get in touch at [embedded-link@codat.io](mailto:embedded-link@codat.io) or use our <a href="https://portal.productboard.com/codat/12-public-devex-roadmap/c/485-embed-a-pre-built-auth-flow-in-your-website-or-app" target="_blank">public roadmap</a> to request new features and enhancements.

:::

## What's New, June 2023

The June 2023 release of Embedded Link brings the following enhancements:

> * **Support for non-React JavaScript apps.** Without a dependency on React, you can use Embedded Link with all JavaScript frameworks, or even with vanilla JavaScript.

Author: Removing the bullet points seemed to work.

* **Increased display control.** Now, you specify dimensions for the Embedded Link component, which will expand to 100% of the given container size. Previously, the component used a fixed width and height.

> * **Navigation improvements.** Source types (accounting, commerce, and banking) can now be connected in any order you choose.

Reviewer 3: There's also file upload as an integration source. Also might be worth noting that we've added indicators of connection success/failure. Any errors are not disruptive to the flow,
![image](https://github.com/codatio/codat-docs/assets/73177718/bc9065ce-74bb-4bc6-bd43-90cabb730837)
![image](https://github.com/codatio/codat-docs/assets/73177718/846aecaf-ba1f-414f-83f5-ab57f15a5499)
![image](https://github.com/codatio/codat-docs/assets/73177718/a6e1bd50-4b44-49b6-9010-4ae0e3d5434e)

Author: Yes, this is definitely worth adding to the list of new features!

* **Performance improvements.** Link now loads quicker and can be loaded only when required.

## Embedded Link overview

Use the Embedded Link solution to benefit from a pre-built code component that melds best practices together with our extensive experience in building authorization flows, while seamlessly embedding it into your webpage or front-end application.

> Embedded Link is a JavaScript component that neatly sits in your front-end code, and can be deployed in a matter of minutes. It supports multiple JavaScript frameworks, including React and TypeScript.

Reviewer 3: Typescript is a programming language, and not a framework

![](/img/link/0014-embedded-link-demo.png)

:::caution Embedded Link is in beta

Embedded Link is in beta, with more enhancements coming soon. You can report any issues with the component by contacting [Codat Support](mailto:support@codat.io).

If you have general feedback on the component, get in touch at [embedded-link@codat.io](mailto:embedded-link@codat.io) or use our <a href="https://portal.productboard.com/codat/12-public-devex-roadmap/c/485-embed-a-pre-built-auth-flow-in-your-website-or-app" target="_blank">public roadmap</a> to request new features and enhancements.

:::

## What's New

The June 2023 release of Embedded Link brings the following enhancements:

> **Support for non-React JavaScript apps**   

Reviewer: Could be #### ?

## Features

* An authorization flow UI built with our expertise and best practices in mind
* Tailored to be flexible and support your custom onboarding and connection journey flows
* Compatible with both React and non-React JavaScript frameworks
* Fast and easy implementation with a pre-built code component
* Authentication in line with OAuth 2.0 standards

> * UI that can be [customized](/auth-flow/customize/customize-link) and reflects your [company branding](/auth-flow/customize/branding)

Reviewer 4: ```suggestion
* UI that can be [customized](/auth-flow/customize/customize-link) and reflects your [company branding](/auth-flow/customize/branding).
```

## Resources

> Our [sample project](https://github.com/codatio/sample-project-link-sdk) shows you how to add Link to a default React project.

Author: The npm library is gone for now (I should have explained this in the PR description).

I should make the call to action clearer.

## Prerequisites

> If you're using React, you'll need to use React version 18 when embedding the Link component into your webpage or application. The component is also backwards compatible with React version 17.

Reviewer 4: These sentences together don't really make sense. If you need to have version 18 then surely it won't work with version 17? 
```suggestion
If you're using React, you'll need to use React version 17+ when embedding the Link component into your webpage or application.
```

> If you haven't already done so,  customize Link on the <a href="https://app.codat.io/settings/link-settings" target="_blank">**Link settings**</a> page in the Codat Portal. For example, add UI copy, set file upload options, and choose whether steps in the Link flow are optional. These settings are applied to Embedded Link as well as Hosted Link.

Author: Thanks. I have added that information.

## Get started with non-React frameworks

> Embedded Link is published to `https://link-sdk.codat.io` as an ES6 module. To use the Embedded Link component in your app with a non-React JavaScript framework, perform the following steps:

Author: One for @lucy-m

> Embedded Link is published to `https://link-sdk.codat.io` as an ES6 module. To use the Embedded Link component in your app with a non-React JavaScript framework, perform the following steps:

Author: Decided to make the link clickable.

1. Using the [Create company](/codat-api#/operations/create-company) endpoint, create a company, and retain its `companyID`. The component needs the `companyId` parameter to open Link for a specified company. For testing purposes, you can create a company in the Codat Portal instead.

   :::note
   We recommend you create a company at the same time as your SMB customer signs up within your app.
   :::
   
1. Create a target `div` for the `CodatLink` component that has the following attributes:
   * An ID of `codat-link-container`.
   * A width and height&mdash;we recommend 460 by 840px.
   The created `CodatLink` component will expand to 100% of the specified dimensions of the `div` container. 
1. Import the component. If you're using the component inside a `script` tag, the tag must have `type="module" import { CodatLink } from "https://link-sdk.codat.io";` set.
1. Initialize the Codat Link component in your app and supply the `companyId` you created in step one:

   ```js Title="Initialize Codat Link component (non-React)"
   const target = document.querySelector("#codat-link-container");
   new CodatLink({
    target,
    props: {
    companyId: "<CODAT_COMPANY_ID>",

>     onClose: () => alert("onClose"),

Author: Please can we sync up on this once you've run through all the steps.

>     onClose: () => alert("onClose"),

Reviewer 3: I'll set some time aside today to run through the steps:)

    onConnection: () => alert("onConnection"),
    onFinish: () => alert("onFinish"),
    onError: () => alert("onError"),
    },
   });
   ```
   
1. If you have a `browserslist` entry in your `package.json` file, make sure your production `browserslist` contains the following entries:

   ```js
    "production": [
      ">0.2% and supports es6-module",
      "not dead",
      "not and_uc >= 0"
    ],
   ```
   
1. If you're using content security policy (CSP) headers, you *must* edit the headers as follows.
   * Add `*.codat.io` to all of `(script-src, style-src, font-src, connect-src, img-src)`, or to `default-src`.
   * Add `unsafe-inline` to `style-src`. Do *not* use a hash because this can change at any time without warning.
1. If you're using TypeScript,  you can use the `types.d.ts` type declaration file. The contents of this file can be added to a new or an existing `.d.ts` file.

> ## Get started with React or TypeScript

Reviewer 3: Typescript and React are not exclusive. You can use React with TS, Javascript with TS, Svelte with TS etc.

Author: So, I changed this heading to "Get started with React".

To use the Embedded Link component in your app with either React or TypeScript, perform the following steps:

> 1. Create a company and retain its company ID (for instructions, see step one of the preceding task).

Author: I don't really want to link at such a granular level. For this step, I think it's worth just repeating it in full.

>    :::note

Reviewer 2: same, recommending a title for this note

   We recommend you create a company at the same time as your SMB customer signs up within your app.
   :::
   

> 1. Copy and paste the contents of the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/d4cde71f-a328-44b7-a573-ed5ad9adc6bb?fileName=types.d.ts&download=true" target="_blank"> `types.d.ts`</a> file into new or existing `.d.ts` file.

Reviewer 2: would the reader be able to download from this link? it's not behind a password or anything?

> 1. Copy and paste the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/11520cd5-799f-4fc2-9497-48d5b509a471?fileName=CodatLink.tsx&download=true" target="_blank">`CodatLink.tsx`</a> file to an appropriate location in your React or TypeScript app. Initialize this component where required.

Reviewer 2: same question

> 1. If you have a `browserslist` entry in your `package.json` file, make sure it's updated for production as shown in step five of the preceding task.

Reviewer 2: more of a discussion point, but perhaps we should be duplicating content here for steps 4 and 5. the user will go through either task 1 or task 2, not both together - so if I want to get started with React, then I will have to read how to get started with non-react, because my section does not give me all the info

Reviewer 3: Agreed! Maybe adding a link to the relevant steps will help?

Author: Good point. I still need to address this one.

> 1. If you're using content security policy (CSP) headers, update them with the information shown in step six of the preceding task.

Reviewer 2: same as above

   ```js
    "production": [
      ">0.2% and supports es6-module",
      "not dead",
      "not and_uc >= 0"
    ],
   ```

1. If you're using content security policy (CSP) headers, you *must* edit the headers as follows.
   * Add `*.codat.io` to all of `(script-src, style-src, font-src, connect-src, img-src)`, or to `default-src`.
   * Add `unsafe-inline` to `style-src`. Do *not* use a hash because this can change at any time without warning.
1. If you're using TypeScript,  you can use the `types.d.ts` type declaration file. The contents of this file can be added to a new or existing `.d.ts` file.

## Get started with React

To use the Embedded Link component in your webpage or app with React, perform the following steps:

1. Using the [Create company](/codat-api#/operations/create-company) endpoint, create a company, and retain its `companyID`. The component needs the `companyId` parameter to open Link for a specified company. You can also create a company in the Codat Portal.

   :::note Company creation timeline
   We recommend you create a company when your SMB customer signs up within your app.
   :::
   

> 1. Download the <a href="https://github.com/codatio/sdk-link/blob/main/snippets/types.d.ts" target="_blank"> `types.d.ts`</a> file, then copy and paste its contents into a new or existing `.d.ts` file.

Reviewer: ```suggestion
1. **Extend your type declarations with our types** - download the <a href="https://github.com/codatio/sdk-link/blob/main/snippets/types.d.ts" target="_blank"> `types.d.ts`</a> file, then copy and paste its contents into a new or existing `.d.ts` file.
```

> 1. Download the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/d4cde71f-a328-44b7-a573-ed5ad9adc6bb?fileName=types.d.ts&download=true" target="_blank"> `types.d.ts`</a> file, then copy and paste its contents into a new or existing `.d.ts` file.

Reviewer: ```suggestion
1. Download the <a href="https://github.com/codatio/sdk-link/blob/main/snippets/types.d.ts" target="_blank"> `types.d.ts`</a> file, then copy and paste its contents into a new or existing `.d.ts` file.
```

> 1. Copy and paste the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/11520cd5-799f-4fc2-9497-48d5b509a471?fileName=CodatLink.tsx&download=true" target="_blank">`CodatLink.tsx`</a> file to an appropriate location in your React or TypeScript app. Initialize this component where required.

Reviewer: ```suggestion
2. **Create a component that mounts the SDK** - you can copy and paste the example <a href="https://github.com/codatio/sdk-link/blob/main/snippets/CodatLink.tsx" target="_blank">`CodatLink.tsx`</a> file to an appropriate location in your React or TypeScript app
```

1. If a `browserslist` entry exists in your `package.json` file, update it with the following entries for production:

   ```js
    "production": [
      ">0.2% and supports es6-module",
      "not dead",
      "not and_uc >= 0"
    ],
   ```
   
1. If you're using content security policy (CSP) headers, you *must* edit the headers as follows:
   * Add `*.codat.io` to all of `(script-src, style-src, font-src, connect-src, img-src)`, or to `default-src`.
   * Add `unsafe-inline` to `style-src`. Do *not* use a hash because this can change at any time without warning.

## Embedding the Link application

> (Q - do we need to keep / move the code snippets below?)

Reviewer: ```suggestion
```

## Features

* An authorization flow UI built with our expertise and best practices in mind
* Tailored to be flexible and support your custom onboarding and connection journey flows
* Compatible with both React and non-React JavaScript frameworks
* Fast and easy implementation with a pre-built code component
* Authentication in line with OAuth 2.0 standards
* UI that can be [customized](/auth-flow/customize/customize-link) and reflects your [company branding](/auth-flow/customize/branding)

## Resources

We've provided a [sample GitHub project](https://github.com/codatio/sample-project-link-sdk) that illustrates how to add the Embedded Link component to a React project.

## Prerequisites

> If you're using React, you'll need to use React version 17 or above when embedding the Link component into your webpage or application.

Reviewer 5: Not a requirement any more, it should work with any version of React

Author: OK. I removed the whole sentence.

## Get started with non-React frameworks

Embedded Link is published to https://link-sdk.codat.io as an ES6 module. To use the Embedded Link component in your app with a non-React JavaScript framework, perform the following steps:

1. Using the [Create company](/codat-api#/operations/create-company) endpoint, create a company, and retain its `companyID`. The component needs the `companyId` parameter to open Link for a specified company. You can also create a company in the Codat Portal.

   :::note Company creation timeline
   We recommend you create a company at the same time as your SMB customer signs up within your app.
   :::
   
1. Create a target `div` for the `CodatLink` component that has the following attributes:
   * An ID of `codat-link-container`.
   * A width and a height&mdash;we recommend 460 by 840px.
   The created `CodatLink` component will expand to 100% of the specified dimensions of the `div` container. 

> 1. Import the component. If you're using the component inside a `script` tag, the tag must have `type="module" set.

Reviewer 3: @lucy-m Would it make sense to change this to "Install the Codat library and import the Link SDK component: import { CodatLink } from "@codat/link-sdk"`;"?

Reviewer 5: The idea of installing doesn't really apply here and is instead being loaded from our servers. I would say:
> Imort the link SDK component: `import { CodatLink} from https://link-sdk.codat.io/`"

> 1. Import the component. If you're using the component inside a `script` tag, the tag must have `type="module" import { CodatLink } from "https://link-sdk.codat.io";` set.

Reviewer 5: ```suggestion
1. Import the component. If you're using the component inside a `script` tag, the tag must have `type="module" set.
```

1. Initialize the Codat Link component in your app and supply the `companyId` you created in step one:

   ```js Title="Initialize Codat Link component (non-React)"
   const target = document.querySelector("#codat-link-container");
   new CodatLink({
    target,
    props: {
    companyId: "<CODAT_COMPANY_ID>",
    onClose: () => alert("onClose"),
    onConnection: () => alert("onConnection"),
    onFinish: () => alert("onFinish"),
    onError: () => alert("onError"),
    },
   });
   ```
   
1. If you have a `browserslist` entry in your `package.json` file, make sure your production `browserslist` contains the following entries:

   ```js
    "production": [
      ">0.2% and supports es6-module",
      "not dead",
      "not and_uc >= 0"
    ],
   ```
   
1. If you're using content security policy (CSP) headers, you *must* edit the headers as follows.
   * Add `*.codat.io` to all of `(script-src, style-src, font-src, connect-src, img-src)`, or to `default-src`.
   * Add `unsafe-inline` to `style-src`. Do *not* use a hash because this can change at any time without warning.

> 1. If you're using TypeScript,  you can use the `types.d.ts` type declaration file. The contents of this file can be added to a new or an existing `.d.ts` file.

Reviewer 5: >  you can use the `types.d.ts` type declaration 

Is there a link to the types.d.ts file anywhere?

## Get started with React

To use the Embedded Link component in your app with React, perform the following steps:

1. Using the [Create company](/codat-api#/operations/create-company) endpoint, create a company, and retain its `companyID`. The component needs the `companyId` parameter to open Link for a specified company. You can also create a company in the Codat Portal.

   :::note Company creation timeline
   We recommend you create a company at the same time as your SMB customer signs up within your app.
   :::
   

> 1. Download the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/d4cde71f-a328-44b7-a573-ed5ad9adc6bb?fileName=types.d.ts&download=true" target="_blank"> `types.d.ts`</a> file, then copy and paste its contents into a new or existing `.d.ts` file.

Reviewer 5: The linked file will not be accessible to people outside of Codat. We will need to host it somewhere and link it here.

Author: Thanks Lucy. I will discuss the hosting options offline and then update.

> 1. Copy and paste the <a href="https://dev.azure.com/codat/ea17b8fb-0083-4bb3-bcac-8a817722f00a/_apis/wit/attachments/11520cd5-799f-4fc2-9497-48d5b509a471?fileName=CodatLink.tsx&download=true" target="_blank">`CodatLink.tsx`</a> file to an appropriate location in your React or TypeScript app. Initialize this component where required.

Reviewer 5: As above, this linked file will not be accessible to people outside of Codat.
